packages/flutter/lib/src/services/hardware_keyboard.dart

```diff
@@ -0,0 +1,894 @@
+  LogicalKeyboardKey? lookUpLayout(PhysicalKeyboardKey physicalKey) => _pressedKeys[physicalKey];
+
+  final Set<KeyboardLockMode> _lockModes = <KeyboardLockMode>{};
+  /// The set of [KeyboardLockMode] that are enabled.
+  ///
+  /// Lock keys, such as CapsLock, are logical keys that toggle their
+  /// respective boolean states on key down events. Such flags are usually used
+  /// as modifier to other keys or events.
+  ///
+  /// If used during a key event handler, the result will have included effect
+  /// of the event.
+  Set<KeyboardLockMode> get lockModesEnabled => _lockModes;
+
+  void _assertEventIsRegular(KeyEvent event) {
+    assert(() {
+      const String common = 'If this occurs in real application, please report this '
+        'bug to Flutter. If this occurs in unit tests, please ensure that '
+        "simulated events follow Flutter's event model as documented in "
+        '`HardwareKeyboard`. This was the event: ';
+      if (event is KeyDownEvent) {
+        assert(!_pressedKeys.containsKey(event.physicalKey),
+          'A ${event.runtimeType} is dispatched, but the state shows that the physical '
+          'key is already pressed. $common$event');
+      } else if (event is KeyRepeatEvent || event is KeyUpEvent) {
+        assert(_pressedKeys.containsKey(event.physicalKey),
+          'A ${event.runtimeType} is dispatched, but the state shows that the physical '
+          'key is not pressed. $common$event');
+        assert(_pressedKeys[event.physicalKey] == event.logicalKey,
+          'A ${event.runtimeType} is dispatched, but the state shows that the physical '
+          'key is pressed on a different logical key. $common$event '
+          'and the recorded logical key ${_pressedKeys[event.physicalKey]}');
+      } else {
+        assert(false, 'Unexpected key event class ${event.runtimeType}');
+      }
+      return true;
+    }());
+  }
+
+  List<KeyEventCallback> _handlers = <KeyEventCallback>[];
+  bool _duringDispatch = false;
+  List<KeyEventCallback>? _modifiedHandlers;
+
+  /// Register a listener that is called every time a hardware key event
+  /// occurs.
+  ///
+  /// All registered handlers will be invoked in order regardless of
+  /// their return value. The return value indicates whether Flutter
+  /// "handles" the event. If any handler returns true, the event
+  /// will not be propagated to other native components in the add-to-app
+  /// scenario.
+  ///
+  /// If used during event dispatching, the addition will not take effect
+  /// until after the dispatching.
+  ///
+  /// See also:
+  ///
+  ///  * [removeHandler], which removes the handler.
+  void addHandler(KeyEventCallback handler) {
+    if (_duringDispatch) {
+      _modifiedHandlers ??= <KeyEventCallback>[..._handlers];
+      _modifiedHandlers!.add(handler);
+    } else {
+      _handlers.add(handler);
+    }
+  }
+
+  /// Stop calling the given listener every time a hardware key event
```

Comment:
You might mention that handlers that are being disposed should remove their handlers.